In [1]:
f = open('day13_input.txt')
lines = f.readlines()
f.close()
data = [(line[0:-1]) for line in lines]

In [2]:
print('data:')
print(data)

data:
['1007125', '13,x,x,41,x,x,x,x,x,x,x,x,x,569,x,29,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,x,19,x,x,x,23,x,x,x,x,x,x,x,937,x,x,x,x,x,37,x,x,x,x,x,x,x,x,x,x,17']


The first line of the data, '1007125' is a timestamp (started from a certain time t=0), and the second line includes the bus IDs (13, 41, 569, 29...).

Each bus departs at t=0, then departs at t=N*id:

bus 13 departs at t=0, 13, 26, 39 ...

bus 41 departs at t=0, 41, 82, 123 ...

...

# Part 1

The question for part 1 is, what the next bus after the given timestamp (1007125)?

In [3]:
t = int(data[0])
print(f'The timestamp is {t}.')
bus_ids = [int(num) for num in data[1].split(',') if num.isnumeric()]
print('The bus ids:')
bus_ids

The timestamp is 1007125.
The bus ids:


[13, 41, 569, 29, 19, 23, 937, 37, 17]

In [4]:
min_wait_t = t
tmp_bus_id = None
for num in bus_ids:
    wait_time = num-t%num
    if min_wait_t > wait_time:
        min_wait_t = wait_time
        tmp_bus_id = num
    print(f'For bus {num}, the wait time is {num-t%num}.')
print('-------')
print(f'The min wait time is {min_wait_t}, for bus {tmp_bus_id}.')

For bus 13, the wait time is 11.
For bus 41, the wait time is 40.
For bus 569, the wait time is 5.
For bus 29, the wait time is 16.
For bus 19, the wait time is 8.
For bus 23, the wait time is 22.
For bus 937, the wait time is 150.
For bus 37, the wait time is 15.
For bus 17, the wait time is 6.
-------
The min wait time is 5, for bus 569.


569x5 will give the answer for part 1:

In [5]:
569 * 5

2845

# Part 2

For this part, we are required to find a time t, when the first list bus will depart, and followed by each subsequent listed bus. In other words, at time t, bus 13 will depart, then bus 41 at a t+3 (because there are 2 x's between 41 and 13), then bus 569 at t+13, and so on... just like they are listed in the input, and x's are important now.

In [6]:
buses = [int(bus) if bus != "x" else bus for bus in data[1].split(",")]
print(buses)

[13, 'x', 'x', 41, 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 569, 'x', 29, 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 19, 'x', 'x', 'x', 23, 'x', 'x', 'x', 'x', 'x', 'x', 'x', 937, 'x', 'x', 'x', 'x', 'x', 37, 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 17]


We need to find a t so that

(t + i) % bus == 0 for i, bus in enumerate(buses), for all the non-x buses.

We are also told that t will be larger than 100000000000000.

In [7]:
t = 100000000000000
while not all([(t + i) % bus == 0 for i, bus in enumerate(buses) if bus != 'x']):
    tmp_buses = [bus for i, bus in enumerate(buses) if bus != 'x' and (t + i)%bus == 0]
    inc = 1
    for bus in tmp_buses:
        inc *= bus       # p is the how much t will increase each step
    t += inc

print(f'The earliest t would be {t}.')

The earliest t would be 487905974205117.
